### PR TITLE
Set ssl_mode for Oracle MySQL if applicable.

### PR DESCRIPTION
**Pull request by dveeden**
Supersedes #82

### CODE REVIEW

**Comment by pali** [comment]
With some versions of mysql client compilation failed... Looks like these continuous-integration tests are really good for catching problems with code which has lot of #ifdefs...

**Comment by pali** [comment]
Anyway, I do not think that usage of SSL/TLS tunel without verifycation of server certificate (either by CA or certificate chain or directly by fingerprint of server certificate) is a good idea. In some cases it could make sense (e.g. passive scanning), but basically it has similar security as not using SSL/TLS tunel.

Without verification of server certificate MITM is possible in same way as MITM for plain text without SSL/TLS.

I really would suggest that verification of server certificate should be preferred option. We probably want to have backward compatibility, but at least examples and documentation can be updated to show people default usage of certificate verification.

In every case it is better if users of DBD::mysql starts doing certificate verification. And documentation or "example" usage could help...

**Comment by dveeden** [comment]
> Anyway, I do not think that usage of SSL/TLS tunel without verifycation of server certificate (either by CA or certificate chain or directly by fingerprint of server certificate) is a good idea. In some cases it could make sense (e.g. passive scanning), but basically it has similar security as not using SSL/TLS tunel.
It is slightly better than not using TLS as it protects against passive attacks.
However it can also give a false sense of security and be a performance overhead.
So no clear winner there.

> Without verification of server certificate MITM is possible in same way as MITM for plain text without SSL/TLS.
Yes indeed.

> I really would suggest that verification of server certificate should be preferred option. We probably want to have backward compatibility, but at least examples and documentation can be updated to show people default usage of certificate verification.
Yes, VERIFY_CA is reasonably secure. VERIFY_IDENTITY is more secure, but hostname validation is in practice not always possible becuase MySQL and MariaDB don't support SubjectAlternativeName.

> In every case it is better if users of DBD::mysql starts doing certificate verification. And documentation or "example" usage could help...
Indeed.

File: dbdimp.c [hunk 1]
```diff
@@ -2083,11 +2081,33 @@ MYSQL *mysql_dr_connect(
 
 	    mysql_ssl_set(sock, client_key, client_cert, ca_file,
 			  ca_path, cipher);
-#if MYSQL_VERSION_ID >= SSL_VERIFY_VERSION && MYSQL_VERSION_ID <= SSL_LAST_VERIFY_VERSION
+#ifdef MYSQL_SSL_MODE
+            if ssl_verify_true:
```
Comment (pali) [inline dbdimp.c]: syntax error. In C it is `if (ssl_verify_true)` and no colon at the end.

File: dbdimp.h [hunk 2]
```diff
@@ -61,12 +61,17 @@
...
+#define SSL_ENFORCE_VERSION 50703
+#define SSL_LAST_ENFORCE_VERSION 50799
 #define MYSQL_VERSION_5_0 50001
+#define MARIADB_VERSION_10 100000
 /* This is to avoid the ugly #ifdef mess in dbdimp.c */
 #if MYSQL_VERSION_ID < SQL_STATE_VERSION
 #define mysql_sqlstate(svsock) (NULL)
 #endif
-
+#ifdef SSL_MODE_PREFERRED
```
Comment (pali) [inline dbdimp.h]: Looks like `SSL_MODE_PREFERRED` is not macro so this #ifdef is always false and `MYSQL_SSL_MODE` is never defined.

File: dbdimp.c [hunk 3]
```diff
@@ -2083,11 +2081,33 @@ MYSQL *mysql_dr_connect(
 
 	    mysql_ssl_set(sock, client_key, client_cert, ca_file,
 			  ca_path, cipher);
-#if MYSQL_VERSION_ID >= SSL_VERIFY_VERSION && MYSQL_VERSION_ID <= SSL_LAST_VERIFY_VERSION
+#ifdef MYSQL_SSL_MODE
+            if (ssl_verify_true)
+              ssl_mode = SSL_MODE_VERIFY_IDENTITY;
+            else if (strlen(ca_file) > 0)
```
Comment (pali) [inline dbdimp.c]: ca_file and also ca_path could be NULL which means strlen here can crash

File: dbdimp.c [hunk 4]
```diff
@@ -2083,11 +2081,33 @@ MYSQL *mysql_dr_connect(
...
+              ssl_mode = SSL_MODE_VERIFY_CA;
+	    mysql_options(sock, MYSQL_OPT_SSL_MODE, &ssl_mode);
+#elif MYSQL_VERSION_ID >= SSL_VERIFY_VERSION && MYSQL_VERSION_ID <= SSL_LAST_VERIFY_VERSION || MYSQL_VERSION_ID >= MARIADB_VERSION_10
 	    mysql_options(sock, MYSQL_OPT_SSL_VERIFY_SERVER_CERT, &ssl_verify_true);
+#if MYSQL_VERSION_ID >= SSL_ENFORCE_VERSION && MYSQL_VERSION_ID <= SSL_LAST_ENFORCE_VERSION
+            /* Only needed for Oracle MySQL 5.7 if MYSQL_OPT_SSL_MODE is not available */
+	    mysql_options(sock, MYSQL_OPT_SSL_ENFORCE, &ssl_verify_true);
+#endif
+#else
+            warn("Can't enable strict certificate checks");
```
Comment (pali) [inline dbdimp.c]: I would say this is a fatal error.

**Comment by pali** [comment]
Now we have there Riddle vulnerability at http://riddle.link/

And looks like Oracle probably fixes SSL enforcement in 5.5 and 5.6 versions of MySQL as documentation was already updated. Compare:
https://web.archive.org/web/20161219180824/https://dev.mysql.com/doc/refman/5.5/en/mysql-ssl-set.html
https://dev.mysql.com/doc/refman/5.5/en/mysql-ssl-set.html

> In MySQL 5.5.55, the MYSQL_OPT_SSL_MODE option for mysql_options() was backported from MySQL 5.7 to MySQL 5.5.

But MySQL 5.5.55 was not released yet...

File: dbdimp.c (hunk 3 above)
Comment (dveeden) [inline dbdimp.c]: Fixed

File: dbdimp.c (hunk 4 above)
Comment (dveeden) [inline dbdimp.c]: Yes I agree. Note that currently there isn't even a warning in this case.

File: dbdimp.c [hunk 5]
```diff
@@ -2083,11 +2081,33 @@ MYSQL *mysql_dr_connect(
...
+              ssl_mode = SSL_MODE_VERIFY_CA;
+	    mysql_options(sock, MYSQL_OPT_SSL_MODE, &ssl_mode);
+#elif MYSQL_VERSION_ID >= SSL_VERIFY_VERSION && MYSQL_VERSION_ID <= SSL_LAST_VERIFY_VERSION || MYSQL_VERSION_ID >= MARIADB_VERSION_10
 	    mysql_options(sock, MYSQL_OPT_SSL_VERIFY_SERVER_CERT, &ssl_verify_true);
+#if MYSQL_VERSION_ID >= SSL_ENFORCE_VERSION && MYSQL_VERSION_ID <= SSL_LAST_ENFORCE_VERSION
+            /* Only needed for Oracle MySQL 5.7 if MYSQL_OPT_SSL_MODE is not available */
+	    mysql_options(sock, MYSQL_OPT_SSL_ENFORCE, &ssl_verify_true);
+#endif
+#else
+            die("Can't enable strict certificate checks");
```
Comment (pali) [inline dbdimp.c]: Use `croak(...)` as you are not passing return value from `die(...)` somewhere else. Look into perlapi for details.

File: dbdimp.c [hunk 6]
```diff
@@ -2083,11 +2081,33 @@ MYSQL *mysql_dr_connect(
...
 			  ca_path, cipher);
-#if MYSQL_VERSION_ID >= SSL_VERIFY_VERSION && MYSQL_VERSION_ID <= SSL_LAST_VERIFY_VERSION
+#ifdef MYSQL_SSL_MODE
+            if (ssl_verify_true)
+              ssl_mode = SSL_MODE_VERIFY_IDENTITY;
+            else if (ca_file && (strlen(ca_file) > 0))
+              ssl_mode = SSL_MODE_VERIFY_CA;
+            else if (ca_path && (strlen(ca_path) > 0))
+              ssl_mode = SSL_MODE_VERIFY_CA;
+	    mysql_options(sock, MYSQL_OPT_SSL_MODE, &ssl_mode);
```
Comment (pali) [inline dbdimp.c]: Seems your patch is whitespace inconsistence with code around... See indentation. If it is possible try to fix it...

File: dbdimp.c [hunk 7]
```diff
@@ -2083,11 +2081,33 @@ MYSQL *mysql_dr_connect(
 
 	    mysql_ssl_set(sock, client_key, client_cert, ca_file,
 			  ca_path, cipher);
-#if MYSQL_VERSION_ID >= SSL_VERIFY_VERSION && MYSQL_VERSION_ID <= SSL_LAST_VERIFY_VERSION
+#ifdef MYSQL_SSL_MODE
+            if (ssl_verify_true)
+              ssl_mode = SSL_MODE_VERIFY_IDENTITY;
+            else if (ca_file && (strlen(ca_file) > 0))
```
Comment (pali) [inline dbdimp.c]: Do we need check for `strlen()`? I have not find any information about it and variables are already passed to `mysql_ssl_set()` without `strlen()` check... So I'm not sure there...

**Comment by dveeden** [comment]
> This code starts to be a big #if mess with long conditions after #if... I do not have idea how to improve it right now.

I'm afraid this will get worse over time because MySQL and MariaDB are (very) slowly diverging. We can try to prevent that as much as possible, but eventually it might make sense to split this into DBD::mysql and DBD::mariadb.

**Comment by pali** [comment]
> I'm afraid this will get worse over time because MySQL and MariaDB are (very) slowly diverging.

Probably...

> We can try to prevent that as much as possible, but eventually it might make sense to split this into DBD::mysql and DBD::mariadb.

But here we have another problem unrelated to MariaDB... Problem is with different MySQL versions. So splitting into DBD::mysql and DBD::mariadb does not help there. And there are still very large common codebase so I do not think split is good idea.

File: dbdimp.c (hunk 7 above)
Comment (dveeden) [inline dbdimp.c]: Removed strlen.

File: dbdimp.c (hunk 6 above)
Comment (dveeden) [inline dbdimp.c]: I've tried to fix it.

File: dbdimp.c (hunk 5 above)
Comment (dveeden) [inline dbdimp.c]: I've fixed that

**Comment by dveeden** [comment]
If `mysql_ssl=1` is not present in the dsn but something like `mysql_ssl_ca_file` is then the latter is ignored. I think we should generate a warning in that case. Not sure what the best way is to do that and I'd like to fix that in a separate pull request.

**Comment by pali** [comment]
I did some SSL changes here: https://github.com/perl5-dbi/DBD-mysql/pull/114 I tried to make that ifdef hell more readable and prepare ground for new MySQL 5.5.55 which should finally have fixed Riddle vulnerability.

**Comment by dveeden** [comment]
@pali should we wait on your ssl branch to be ready to merge? Or should this branch be merged?
I'd like to see something merged to fix the insecure code (not related to Riddle).

**Comment by pali** [comment]
For me this is PR is ready to merge, I already marked it.

**Comment by pali** [comment]
I updated my https://github.com/perl5-dbi/DBD-mysql/pull/114 on top of master (so including this PR).

**Comment by pali** [comment]
@dveeden this PR #108 broke compilation for MySQL C/Connector 6.0:
https://travis-ci.org/perl5-dbi/DBD-mysql/builds/218257554

**Pull request merged**